fix: add --require-citations and --require-tests validation flags

seebees opened the pull request:
Add optional validation flags to enforce citation and test coverage:
- --require-citations: Fails if spec requirements lack implementation citations
- --require-tests: Fails if citations lack corresponding test annotations

Features:
- Both flags support boolean values (--flag, --flag=true, --flag=false)
- Default to false for backward compatibility
- Provide clear error messages when validation fails
- Work independently or together for comprehensive validation

Tests:
- Unit tests for CLI argument parsing and validation logic
- Integration tests covering success/failure scenarios for both flags
- Snapshot tests ensuring consistent error message formatting


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

seebees committed:
fix: add --require-citations and --require-tests validation flags

Add optional validation flags to enforce citation and test coverage:
- --require-citations: Fails if spec requirements lack implementation citations
- --require-tests: Fails if citations lack corresponding test annotations

Features:
- Both flags support boolean values (--flag, --flag=true, --flag=false)
- Default to false for backward compatibility
- Provide clear error messages when validation fails
- Work independently or together for comprehensive validation

Tests:
- Unit tests for CLI argument parsing and validation logic
- Integration tests covering success/failure scenarios for both flags
- Snapshot tests ensuring consistent error message formatting

seebees commented on integration/snapshots/report-require-citations-missing_stderr.snap:
```diff
@@ -0,0 +1,29 @@
+       Wrote duvet_report.html 
+     Writing duvet_snapshot.txt
+       Wrote duvet_snapshot.txt 
+  × Specification requirements missing citation.
```
Sure, love to.
On that note, would you like to change it to `--require-implementation`? Or should I add that one as well?

camshaft replied:
Yeah i'd really like to use the `implementation` phrasing over `citation` since that's how it ends up getting used.

seebees commented on duvet/src/report.rs:
```diff
@@ -37,14 +39,14 @@ pub struct Report {
+    require_citations: Option<bool>,
 
-    #[clap(long)]
-    require_tests: Option<Option<bool>>,
+    #[clap(long, action = clap::ArgAction::Set, num_args = 0..=1, default_missing_value = "true")]
+    require_tests: Option<bool>,
 
-    #[clap(long)]
-    ci: Option<Option<bool>>,
+    #[clap(long, action = clap::ArgAction::SetTrue)]
+    ci: bool,
```
I don't remember. I don't know why they were broken up.
I think that there is even some ci env logic.

seebees replied:
Take a look at all the stuff I added. We can add these options to the config. But I would suggest that you let it still be run from the CLI. With the new filtering logic I have found it to be very helpful.

seebees replied:
Another option would be to make a `audit` or `check` command that does this.
This command would then be configurable in `.duvet/config.toml` and with a few command lines. This could also be the entry point for a code coverage check like what we were talking about offline.

seebees committed:
feat: enhance error output to list annotations

- Replace generic error messages with detailed, actionable feedback
- Categorize errors into missing citations, TODO annotations, and missing tests
- Output complete duvet annotations that developers can copy-paste directly
- Show full requirement text without truncation for better context
- Add comprehensive test coverage for enhanced error formatting

seebees committed:
feat: Targeted requirements  --require* options

Add support for specifying targeted requirements that apply only to specific
specifications or sections, while maintaining backward compatibility with the
existing boolean format.

Changes:
- Support multiple --require-citations and --require-tests flags
- Add spec path targeting (e.g., spec.md, spec.md#section-name)
- Implement exact section name matching for targeted requirements
- Add validation to prevent mixing boolean and path formats
- Update validation logic to apply requirements selectively
- Maintain backward compatibility with existing boolean usage

Examples:
  duvet report --require-citations true  # Old format (still works)
  duvet report --require-citations spec.md  # New: target entire spec
  duvet report --require-citations spec.md#section  # New: target specific section

The parser now supports the exact usage pattern requested:
  duvet report \
    --require-citations AwsLockbox/specification/epoch-manager.md \
    --require-citations AwsLockbox/specification/esdk/esdk-epoch-enclosure.md#createepochkey-security-requirements

This enables more granular control over which parts of specifications require
citations and tests, improving the tool's flexibility for large projects with
multiple specification documents.

seebees committed:
cargo fmt
